drivers/usb/udc/udc_stm32.c

```diff
@@ -0,0 +1,1155 @@
+				struct net_buf *buf)
+{
+	unsigned int lock_key;
+	int ret;
+
+	udc_buf_put(epcfg, buf);
+
+	lock_key = irq_lock();
+
+	if (USB_EP_DIR_IS_IN(epcfg->addr)) {
+		ret = udc_stm32_tx(dev, epcfg->addr, buf);
+	} else {
+		ret = udc_stm32_rx(dev, epcfg->addr, buf);
+	}
+
+	irq_unlock(lock_key);
+
+	return ret;
+}
+
+static int udc_stm32_ep_dequeue(const struct device *dev,
+				struct udc_ep_config *epcfg)
+{
+	struct net_buf *buf;
+
+	udc_stm32_ep_flush(dev, epcfg);
+
+	buf = udc_buf_get_all(dev, epcfg->addr);
+	if (buf) {
+		udc_submit_event(dev, UDC_EVT_EP_REQUEST, -ECONNABORTED, buf);
+	}
+
+	udc_ep_set_busy(dev, epcfg->addr, false);
+
+	return 0;
+}
+
+static const struct udc_api udc_stm32_api = {
+	.lock = udc_stm32_lock,
+	.unlock = udc_stm32_unlock,
+	.init = udc_stm32_init,
+	.enable = udc_stm32_enable,
+	.disable = udc_stm32_disable,
+	.shutdown = udc_stm32_shutdown,
+	.set_address = udc_stm32_set_address,
+	.host_wakeup = udc_stm32_host_wakeup,
+	.ep_try_config = NULL,
+	.ep_enable = udc_stm32_ep_enable,
+	.ep_disable = udc_stm32_ep_disable,
+	.ep_set_halt = udc_stm32_ep_set_halt,
+	.ep_clear_halt = udc_stm32_ep_clear_halt,
+	.ep_enqueue = udc_stm32_ep_enqueue,
+	.ep_dequeue = udc_stm32_ep_dequeue,
+};
+
+/* ----------------- Instance/Device specific data ----------------- */
+
+/*
+ * USB, USB_OTG_FS and USB_DRD_FS are defined in STM32Cube HAL and allows to
+ * distinguish between two kind of USB DC. STM32 F0, F3, L0 and G4 series
+ * support USB device controller. STM32 F4 and F7 series support USB_OTG_FS
+ * device controller. STM32 F1 and L4 series support either USB or USB_OTG_FS
+ * device controller.STM32 G0 series supports USB_DRD_FS device controller.
+ *
+ * WARNING: Don't mix USB defined in STM32Cube HAL and CONFIG_USB_* from Zephyr
+ * Kconfig system.
+ */
+#define USB_NUM_BIDIR_ENDPOINTS	DT_INST_PROP(0, num_bidir_endpoints)
+
+#if defined(USB) || defined(USB_DRD_FS)
+#define EP0_MPS PCD_EP0MPS_64
+#define EP_MPS 64U
+#define USB_BTABLE_SIZE  (8 * USB_NUM_BIDIR_ENDPOINTS)
+#define USB_RAM_SIZE	0
+#else /* USB_OTG_FS */
+#define EP0_MPS USB_OTG_MAX_EP0_SIZE
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs)
+#define EP_MPS USB_OTG_HS_MAX_PACKET_SIZE
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otgfs) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usb)
+#define EP_MPS USB_OTG_FS_MAX_PACKET_SIZE
+#endif
+#define USB_RAM_SIZE	DT_INST_PROP(0, ram_size)
+#define USB_BTABLE_SIZE 0
+#endif /* USB */
+
+#define USB_OTG_HS_EMB_PHY (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc) && \
+			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+
+#define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
+			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+
+static struct udc_stm32_data udc0_priv;
+
+static struct udc_data udc0_data = {
+	.mutex = Z_MUTEX_INITIALIZER(udc0_data.mutex),
+	.priv = &udc0_priv,
+};
+
+static const struct udc_stm32_config udc0_cfg  = {
+	.num_endpoints = USB_NUM_BIDIR_ENDPOINTS,
+	.dram_size = USB_RAM_SIZE,
```

Comment:
On nucleo_wb55rg, I'm getting error:
```
/path/zephyr/drivers/usb/udc/udc_stm32.c: In function 'udc_stm32_ep_mem_config':
/path/zephyr/drivers/usb/udc/udc_stm32.c:412:46: error: 'struct udc_stm32_data' has no member named 'dram_size'
  412 |         if (priv->occupied_mem + size >= priv->dram_size) {
      |                                              ^~
```
It seems that case for this series is not handled correctly. Would you mind having a look ?

Reply:
@erwango should be fixed now.

Reply:
@loicpoulain Compilation is ok, but I'm getting following error:
```
*** Booting Zephyr OS build zephyr-v3.3.0-838-g6d7e63f12d40 ***
[00:00:00.000,000] <err> udc_stm32: Unable to allocate FIFO for 0x00
[00:00:00.000,000] <err> udc_stm32: Failed enabling ep 0x00
[00:00:00.000,000] <err> usbd_dev: Failed to enable controller
[00:00:00.000,000] <err> cdc_acm_echo: Failed to enable device support
[00:00:00.000,000] <err> cdc_acm_echo: Failed to enable USB
```
On a nucleo_f767zi/b_l4s5i_iot01a, I'm getting the following, which looks ok:
```
*** Booting Zephyr OS build zephyr-v3.3.0-838-g6d7e63f12d40 ***
[00:00:00.001,000] <inf> cdc_acm_echo: Wait for DTR
[00:00:06.103,000] <wrn> usbd_core: RESET event
[00:00:06.103,000] <wrn> usbd_core: Bus reset event
[00:00:06.231,000] <wrn> usbd_core: RESET event
[00:00:06.231,000] <wrn> usbd_core: Bus reset event
[00:00:06.330,000] <inf> usbd_cdc_acm: Configuration enabled

```

Reply:
Ok, this 'should' be fixed now, though this board/code-path is blind change for me, as I do not have same MCU based board.

Reply:
Problem is still present unfortunately. Here are some traces:
```
*** Booting Zephyr OS build zephyr-v3.3.0-838-ge1ba4f8dd554 ***
[00:00:00.000,000] <dbg> udc_stm32: udc_stm32_init: 
[00:00:00.000,000] <dbg> udc_stm32: __clock_enable: 
[00:00:00.000,000] <dbg> udc: ep_check_config: cfg d:I|- t:S|B|I|-, mps 1023
[00:00:00.000,000] <dbg> udc: ep_check_config: cfg d:I|- t:S|B|I|-, mps 1023
[00:00:00.000,000] <dbg> udc: ep_check_config: cfg d:-|O t:S|B|I|-, mps 64
[00:00:00.000,000] <dbg> udc_stm32: udc_stm32_enable: 
[00:00:00.000,000] <dbg> udc: ep_check_config: cfg d:-|O t:-|-|-|C, mps 0
[00:00:00.001,000] <dbg> udc_stm32: udc_stm32_ep_enable: 0x00
[00:00:00.001,000] <dbg> udc: ep_check_config: cfg d:I|- t:-|-|-|C, mps 0
[00:00:00.001,000] <dbg> udc_stm32: udc_stm32_ep_enable: 0x80
[00:00:00.001,000] <dbg> udc_stm32: HAL_PCD_ResetCallback: 
[00:00:00.001,000] <wrn> usbd_core: RESET event
[00:00:00.001,000] <wrn> usbd_core: Bus reset event
[00:00:00.001,000] <err> usbd_core: Failed to set default address after bus reset
[00:00:00.001,000] <inf> cdc_acm_echo: Wait for DTR
[00:00:00.003,000] <dbg> udc_stm32: HAL_PCD_SuspendCallback: 
[00:00:00.004,000] <wrn> usbd_core: SUSPEND event
[00:04:25.875,000] <dbg> udc_stm32: HAL_PCD_ResumeCallback: 
[00:04:25.875,000] <wrn> usbd_core: RESUME event
[00:04:26.071,000] <dbg> udc_stm32: HAL_PCD_ResetCallback: 
[00:04:26.071,000] <wrn> usbd_core: RESET event
[00:04:26.071,000] <wrn> usbd_core: Bus reset event
[00:04:26.071,000] <dbg> udc_stm32: udc_stm32_set_address: 0
[00:04:26.071,000] <inf> udc: ep 0x80 is not halted|disabled
[00:04:26.071,000] <dbg> udc: udc_debug_ep_enqueue: ep 0x80 queue is empty
[00:04:26.200,000] <dbg> udc_stm32: HAL_PCD_ResetCallback: 
[00:04:26.200,000] <wrn> usbd_core: RESET event
[00:04:26.200,000] <wrn> usbd_core: Bus reset event
[00:04:26.200,000] <dbg> udc_stm32: udc_stm32_set_address: 0
[00:04:26.200,000] <inf> udc: ep 0x80 is not halted|disabled
[00:04:26.200,000] <dbg> udc: udc_debug_ep_enqueue: ep 0x80 queue is empty
[00:04:26.440,000] <dbg> udc_stm32: HAL_PCD_ResetCallback: 
[00:04:26.440,000] <wrn> usbd_core: RESET event
[00:04:26.440,000] <wrn> usbd_core: Bus reset event
[00:04:26.440,000] <dbg> udc_stm32: udc_stm32_set_address: 0
[00:04:26.440,000] <inf> udc: ep 0x80 is not halted|disabled
[00:04:26.440,000] <dbg> udc: udc_debug_ep_enqueue: ep 0x80 queue is empty
[00:04:26.675,000] <dbg> udc_stm32: HAL_PCD_ResetCallback: 
[00:04:26.675,000] <wrn> usbd_core: RESET event
[00:04:26.675,000] <wrn> usbd_core: Bus reset event
```
Let' me know if this is sufficient to debug, otherwise we'll proceed differently.